Sources/Signal.swift

```diff
@@ -119,20 +119,28 @@ public final class Signal<Value, Error: Swift.Error> {
 				}
 
 				self.sendLock.unlock()
-			}
 
-			// Check if the status has been bumped to `terminating` due to a
-			// terminal event being sent concurrently or recursively.
-			//
-			// The check is deliberately made outside of the `sendLock` so that it
-			// covers also any potential concurrent terminal event in one shot.
-			//
-			// Related PR:
-			// https://github.com/ReactiveCocoa/ReactiveSwift/pull/112
-			if .shouldDispose == self.tryToCommitTermination() {
-				// Dispose only after notifying observers, so disposal
-				// logic is consistently the last thing to run.
-				self.disposable.dispose()
+				// Check if the status has been bumped to `terminating` due to a
+				// terminal event being sent concurrently or recursively.
+				//
+				// The check is deliberately made outside of the `sendLock` so that it
+				// covers also any potential concurrent terminal event in one shot.
+				//
+				// Related PR:
+				// https://github.com/ReactiveCocoa/ReactiveSwift/pull/112
+				//
+				// While calling `tryToCommitTermination` is sufficient, this is a fast
+				// path for the recurring value delivery.
+				//
+				// Note that this cannot be `try` since any concurrent observer bag
+				// manipulation might then cause the terminating state being missed.
+				stateLock.lock()
+				if case .terminating = state {
```

Comment:
This is resurrected because acquiring the spinlock is consistently & considerably faster than calling `tryToCommitTermination()` on my machine (>50ns per call). 🤦‍♂️